src/test/java/com/artipie/asto/lock/RetryLockTest.java

```diff
@@ -0,0 +1,146 @@
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.asto.lock;
+
+import com.google.common.base.Stopwatch;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.core.IsEqual;
+import org.hamcrest.number.IsCloseTo;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.Timeout;
+
+/**
+ * Test cases for {@link RetryLock}.
+ *
+ * @since 0.24
+ * @checkstyle MagicNumberCheck (500 lines)
+ */
+final class RetryLockTest {
+
+    /**
+     * Scheduler used in tests.
+     */
+    private ScheduledExecutorService scheduler;
+
+    @BeforeEach
+    void setUp() {
+        this.scheduler = Executors.newSingleThreadScheduledExecutor();
+    }
+
+    @AfterEach
+    void tearDown() {
+        this.scheduler.shutdown();
+    }
+
+    @Test
+    @Timeout(3)
+    void shouldSucceedAfterSomeAttempts() {
+        final int attempts = 2;
+        final AtomicInteger attempt = new AtomicInteger();
+        new RetryLock(
+            this.scheduler,
+            new Lock() {
+                @Override
+                public CompletionStage<Void> acquire() {
+                    final CompletableFuture<Void> result;
+                    if (attempt.incrementAndGet() < attempts) {
+                        result = new CompletableFuture<>();
+                        result.completeExceptionally(new RuntimeException());
+                    } else {
+                        result = CompletableFuture.allOf();
+                    }
+                    return result;
+                }
+
+                @Override
+                public CompletionStage<Void> release() {
+                    throw new UnsupportedOperationException();
+                }
+            }
+        ).acquire().toCompletableFuture().join();
+        MatcherAssert.assertThat(
+            attempt.get(),
+            new IsEqual<>(attempts)
+        );
+    }
+
+    @Test
+    @Timeout(10)
```

Comment:
@olegmoz can we make shorter interval test timeout? 10 seconds seems to be quite long